src/main/java/com/mongodb/hibernate/internal/translate/AbstractMqlTranslator.java

```diff
@@ -580,17 +577,93 @@ public void visitTuple(SqlTuple sqlTuple) {
 
     @Override
     public void visitDeleteStatement(DeleteStatement deleteStatement) {
-        throw new FeatureNotSupportedException("TODO-HIBERNATE-46 https://jira.mongodb.org/browse/HIBERNATE-46");
+        var pair = getCollectionAstFilterPair(deleteStatement);
+        astVisitorValueHolder.yield(COLLECTION_MUTATION, new AstDeleteCommand(pair.collection, pair.filter));
     }
 
     @Override
     public void visitUpdateStatement(UpdateStatement updateStatement) {
-        throw new FeatureNotSupportedException("TODO-HIBERNATE-46 https://jira.mongodb.org/browse/HIBERNATE-46");
+        var collectionAndFilter = getCollectionAstFilterPair(updateStatement);
+        var fieldUpdates =
+                new ArrayList<AstFieldUpdate>(updateStatement.getAssignments().size());
+        for (var assignment : updateStatement.getAssignments()) {
+            var fieldReferences = assignment.getAssignable().getColumnReferences();
+            if (fieldReferences.size() != 1) {
+                throw new FeatureNotSupportedException();
+            }
+            var fieldPath = fieldReferences.get(0).getColumnReference().getColumnExpression();
+            var assignedValue = assignment.getAssignedValue();
+
+            var sqlTuple = SqlTupleContainer.getSqlTuple(assignedValue);
+            if (sqlTuple != null) {
+                throw new FeatureNotSupportedException();
+            }
+            if (!isValueExpression(assignedValue)) {
+                throw new FeatureNotSupportedException();
+            }
+            var fieldValue = acceptAndYield(assignedValue, FIELD_VALUE);
+            fieldUpdates.add(new AstFieldUpdate(fieldPath, fieldValue));
+        }
+        astVisitorValueHolder.yield(
+                COLLECTION_MUTATION,
+                new AstUpdateCommand(collectionAndFilter.collection, collectionAndFilter.filter, fieldUpdates));
+    }
+
+    private CollectionAstFilterPair getCollectionAstFilterPair(
+            AbstractUpdateOrDeleteStatement updateOrDeleteStatement) {
+        checkMutationStatement(updateOrDeleteStatement);
+        var fromClause = updateOrDeleteStatement.getFromClause();
+        checkFromClauseSupportability(fromClause);
+        var collection = getMutationCollectionAfterRegisteredAsAffectedTable(updateOrDeleteStatement);
+        var filter = acceptAndYield(updateOrDeleteStatement.getRestriction(), FILTER);
+        return new CollectionAstFilterPair(collection, filter);
     }
 
     @Override
-    public void visitInsertStatement(InsertSelectStatement insertSelectStatement) {
-        throw new FeatureNotSupportedException("TODO-HIBERNATE-46 https://jira.mongodb.org/browse/HIBERNATE-46");
+    public void visitInsertStatement(InsertSelectStatement insertStatement) {
+        checkMutationStatement(insertStatement);
+        if (insertStatement.getConflictClause() != null) {
+            throw new FeatureNotSupportedException();
+        }
```

Comment:
HQL has an interesting ON CONFLICT during insertion feature (see https://www.baeldung.com/hibernate-insert-query-on-conflict-clause). It is sort of similar to MongoDB's `upsert` feature but it seems there is a big difference between them:

- HQL mainly focuses on conflict during insertion (w.r.t. unique constraint or fields combination); MongoDB's `upsert` focuses on conflict during updating (w.r.t. some query or filter,, e.g. comparison between a field and a value)

So I guess we can't support this HQL feature. HQL's update statement is simpler and provides no possibility to emulate MongoDB's `upsert` feature.

Reply:
Definitively it deserves a ticket. I think all of the following HQL criteria needs to be satisfied to allow for MQL translation:

- ON CONFLICT is specified
- ON CONSTRAINT uniqueIndexName is not used as the criteria (MQL's filter can't be based on unique index name, correct?)
- CONFLICT ACTION is not NOTHING

so it seems the ticket is an interesting tech one but could be used very rarely in reality; native query seems more flixible and powerful.